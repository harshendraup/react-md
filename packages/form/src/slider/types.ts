import { Dispatch, HTMLAttributes, ReactNode, SetStateAction } from "react";

export type ThumbIndex = 0 | 1;
export type SliderThumbIndex = ThumbIndex | null;
export type SliderDragEvent =
  | MouseEvent
  | TouchEvent
  | React.MouseEvent
  | React.TouchEvent;
export type SliderDraggingBy = "mouse" | "touch" | null;

export type SimpleSliderValue = number;
export type SimpleSliderDefaultValue =
  | SimpleSliderValue
  | (() => SimpleSliderValue);

/**
 * An object containing the functions required to update the `SimpleSlider`'s
 * value when the user interacts with the slider. These functions are provided
 * by the `useSimpleSlider` hook.
 */
export interface SimpleSliderControls {
  minimum(): void;
  maximum(): void;
  increment(): void;
  incrementJump(): void;
  decrement(): void;
  decrementJump(): void;
  setValue: Dispatch<SetStateAction<SimpleSliderValue>>;
}

export type RangeSliderValue = readonly [number, number];
export type RangeSliderDefaultValue =
  | RangeSliderValue
  | (() => RangeSliderValue);

/**
 * An object containing the functions required to update the `RangeSlider`'s
 * value when the user interacts with the slider. These functions are provided by
 * the `useRangeSlider` hook.
 */
export interface RangeSliderControls {
  minimum(index: ThumbIndex): void;
  maximum(index: ThumbIndex): void;
  increment(index: ThumbIndex): void;
  incrementJump(index: ThumbIndex): void;
  decrement(index: ThumbIndex): void;
  decrementJump(index: ThumbIndex): void;
  setValue: Dispatch<SetStateAction<RangeSliderValue>>;
}

export interface SliderValueOptions {
  /**
   * The min value for the slider.
   */
  min?: number;

  /**
   * The max value for the slider.
   */
  max?: number;

  /**
   * A positive number representing the value to "jump" while incrementing or
   * decrementing the slider's value. This should normally stay as the default
   * value of `1`, but can also be decimal values if needed.
   */
  step?: number;
}

export interface SliderStepOptions extends SliderValueOptions {
  /**
   * An optional amount to jump by when using the PageUp or PageDown keys. When
   * this is omitted, it will try to default to 10% of the full range to the
   * nearest step
   */
  jump?: number;
}

export interface SliderPresentation {
  /**
   * Boolean if the slider is rendered vertically instead of horizontally.
   */
  vertical?: boolean;

  /**
   * Boolean if the slider is disabled and the values cannot be changed.
   */
  disabled?: boolean;
}

export interface SliderThumbOptions
  extends Omit<SliderValueOptions, "step">,
    SliderPresentation {
  /**
   * A function that is used to help with accessibility by creating a better
   * value string if just a number isn't representative enough of your range.
   */
  getValueText?(value: number): string;
}

export type DefinedSliderValueOptions = Required<SliderValueOptions>;
export type SliderEventHandlerNames =
  | "onKeyDown"
  | "onMouseDown"
  | "onTouchStart";

export type SliderEventHandlers = Pick<
  HTMLAttributes<HTMLSpanElement>,
  SliderEventHandlerNames
>;

export interface SliderAddons {
  /**
   * An optional addon to render before the slider's track. This can be anything
   * and will be wrapped in the `TextIconSpacing` component from
   * `@react-md/icon`.
   *
   * When the `vertical` prop is enabled, the addon will appear underneath the
   * track.
   */
  beforeAddon?: ReactNode;

  /**
   * An optional addon to render after the slider's track. This can be anything
   * and will be wrapped in the `TextIconSpacing` component from
   * `@react-md/icon`.
   *
   * When the `vertical` prop is enabled, the addon will appear above the track.
   */
  afterAddon?: ReactNode;
}

/**
 * These are the shared and common props required for both the `SimpleSlider` and
 * `RangeSlider` components.
 */
export interface SliderProps
  extends HTMLAttributes<HTMLDivElement>,
    SliderPresentation,
    SliderAddons,
    Pick<SliderThumbOptions, "getValueText"> {
  /**
   * An id for the slider and different parts which is required for a11y.
   */
  baseId: string;

  /**
   * The duration that it takes for the slider animation to complete for a new
   * value. This is just used to help make things look smoother while dragging
   * comapred to jumping to new values.
   */
  animationDuration?: number;

  /**
   * Boolean if the track should inverse the active and inactive states so that
   * the thicker bar would appear on the right instead of the left when
   * horizontal. The thicker bar would appear above instead of below for vertical
   * sliders.
   *
   * Note: This does not do anything for sliders that have two thumbs (range
   * slider).
   */
  inversed?: boolean;
}
