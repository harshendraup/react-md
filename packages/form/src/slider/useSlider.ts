import { useCallback, useMemo, useRef, useState } from "react";
import { nearest } from "@react-md/utils";

import {
  DEFAULT_SLIDER_MAX,
  DEFAULT_SLIDER_MIN,
  DEFAULT_SLIDER_STEP,
} from "./constants";
import {
  DefinedSliderValueOptions,
  SliderControls,
  SliderDefaultValue,
  SliderValue,
  SliderStepOptions,
} from "./types";
import { getJumpValue, getSteps } from "./utils";

/**
 * @since 2.5.0
 */
export interface SliderRequiredProps
  extends SliderControls,
    DefinedSliderValueOptions {
  /**
   * The current value of the slider.
   */
  value: SliderValue;
}

/**
 * @since 2.5.0
 */
export type SliderValueReturnType = readonly [SliderValue, SliderRequiredProps];

/**
 * This hook is used to control the value and behavior of the `Slider`
 * component. The first argument will contain the current slider value while
 * the second argument will be all the props required to control the `Slider`
 * component.
 *
 * @param defaultValue An optional default value to use for the slider. This will
 * default to the `min` option when undefined.
 * @param options An object containing the `min` and `max` values allowed for the
 * slider as well as a `step` to indicate valid values between the `min` and
 * `max`.
 * @return an ordered list containing the current value followed by the `Slider`
 * props
 * @since 2.5.0
 */
export function useSlider(
  defaultValue?: SliderDefaultValue,
  {
    min = DEFAULT_SLIDER_MIN,
    max = DEFAULT_SLIDER_MAX,
    step = DEFAULT_SLIDER_STEP,
    jump: propJump,
  }: SliderStepOptions = {}
): SliderValueReturnType {
  const jump = useMemo(() => getJumpValue(min, max, step, propJump), [
    min,
    max,
    step,
    propJump,
  ]);

  const [value, setValue] = useState(defaultValue ?? min);
  const increment = useCallback(() => {
    setValue((prevValue) => Math.max(min, Math.min(max, prevValue + step)));
  }, [min, max, step]);
  const incrementJump = useCallback(() => {
    setValue((prevValue) => Math.max(min, Math.min(max, prevValue + jump)));
  }, [min, max, jump]);
  const decrement = useCallback(() => {
    setValue((prevValue) => Math.max(min, Math.min(max, prevValue - step)));
  }, [min, max, step]);
  const decrementJump = useCallback(() => {
    setValue((prevValue) => Math.max(min, Math.min(max, prevValue - jump)));
  }, [min, max, jump]);
  const minimum = useCallback(() => {
    setValue(min);
  }, [min]);
  const maximum = useCallback(() => {
    setValue(max);
  }, [max]);

  const prev = useRef({ min, max, step });
  if (
    prev.current.min !== min ||
    prev.current.max !== max ||
    prev.current.step !== step
  ) {
    // ensure that if the `min`, `max`, or `step` value changes that the value
    // is updated as well. Without this, there will be a runtime error if the
    // value is not within the new range.
    prev.current = { min, max, step };
    setValue(nearest(value, min, max, getSteps(min, max, step)));
  }

  return [
    value,
    {
      min,
      max,
      step,
      value,
      increment,
      incrementJump,
      decrement,
      decrementJump,
      minimum,
      maximum,
      setValue,
    },
  ];
}
