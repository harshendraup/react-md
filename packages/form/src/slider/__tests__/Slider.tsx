import React, { ReactElement } from "react";
import { fireEvent, render } from "@testing-library/react";

import { Slider, SliderProps } from "../Slider";
import { SliderStepOptions, SliderValue } from "../types";
import { useSlider } from "../useSlider";

interface TestProps
  extends SliderStepOptions,
    Pick<
      SliderProps,
      | "thumbProps"
      | "onMouseDown"
      | "onTouchStart"
      | "vertical"
      | "disabled"
      | "getValueText"
    > {
  defaultValue?: SliderValue;
}

function Test({
  defaultValue,
  disabled,
  vertical,
  thumbProps,
  onMouseDown,
  onTouchStart,
  getValueText,
  ...options
}: TestProps): ReactElement {
  const [, controls] = useSlider(defaultValue, options);

  return (
    <Slider
      {...controls}
      baseId="slider"
      label="Volume"
      disabled={disabled}
      vertical={vertical}
      thumbProps={thumbProps}
      onMouseDown={onMouseDown}
      onTouchStart={onTouchStart}
      getValueText={getValueText}
    />
  );
}

describe("Slider", () => {
  it("should render correctly", () => {
    const { container, getByRole, rerender } = render(<Test />);
    expect(container).toMatchSnapshot();

    const slider = getByRole("slider");
    slider.focus();
    expect(container).toMatchSnapshot();

    slider.blur();
    expect(container).toMatchSnapshot();

    rerender(<Test vertical />);
    expect(container).toMatchSnapshot();

    slider.focus();
    expect(container).toMatchSnapshot();

    slider.blur();
    expect(container).toMatchSnapshot();

    rerender(<Test disabled />);
    expect(container).toMatchSnapshot();

    rerender(<Test disabled vertical />);
    expect(container).toMatchSnapshot();
  });

  it("should allow for custom valuetext", () => {
    const { getByRole, rerender } = render(
      <Test getValueText={(value) => `$${value}`} />
    );

    const slider = getByRole("slider");
    expect(slider).toHaveAttribute("aria-valuenow", "0");
    expect(slider).toHaveAttribute("aria-valuetext", "$0");

    rerender(<Test getValueText={() => ""} />);
    expect(slider).toHaveAttribute("aria-valuenow", "0");
    expect(slider).not.toHaveAttribute("aria-valuetext");

    rerender(<Test />);
    expect(slider).toHaveAttribute("aria-valuenow", "0");
    expect(slider).not.toHaveAttribute("aria-valuetext");
  });

  it("should call the prop events correctly", () => {
    const onKeyDown = jest.fn();
    const onMouseDown = jest.fn();
    const onTouchStart = jest.fn();

    const { getByRole } = render(
      <Test
        thumbProps={{ onKeyDown }}
        onMouseDown={onMouseDown}
        onTouchStart={onTouchStart}
      />
    );

    const slider = getByRole("slider");

    fireEvent.keyDown(slider);
    expect(onKeyDown).toBeCalledTimes(1);

    fireEvent.mouseDown(slider, { altKey: true });
    expect(onMouseDown).toBeCalledTimes(1);

    fireEvent.touchStart(slider);
    expect(onTouchStart).toBeCalledTimes(1);
  });

  it("should ensure the value stays within the range if it changes after initial render to prevent errors from being thrown", () => {
    const props = {
      defaultValue: 5,
      min: 0,
      max: 5,
      step: 1,
    };
    const { rerender, getByRole } = render(<Test {...props} />);

    const slider = getByRole("slider");
    expect(slider).toHaveAttribute("aria-valuenow", "5");

    rerender(<Test {...props} step={2} max={6} />);
    expect(slider).toHaveAttribute("aria-valuenow", "6");

    rerender(<Test {...props} step={3} max={6} />);
    expect(slider).toHaveAttribute("aria-valuenow", "6");

    rerender(<Test {...props} step={10} max={100} />);
    expect(slider).toHaveAttribute("aria-valuenow", "10");

    rerender(<Test {...props} step={1} min={50} max={100} />);
    expect(slider).toHaveAttribute("aria-valuenow", "50");
  });

  it("should log an error if the slider does not have a valid label", () => {
    function Test(
      props: Pick<
        SliderProps,
        "label" | "thumbLabel" | "thumbLabelledBy" | "thumbProps"
      >
    ): ReactElement {
      const [, controls] = useSlider();

      return <Slider {...props} {...controls} baseId="slider" />;
    }

    const error = jest.spyOn(console, "error").mockImplementation(() => {});
    const { rerender } = render(<Test label="Label" />);
    rerender(<Test thumbLabel="Label" />);
    rerender(<Test thumbLabelledBy="some-id" />);
    rerender(<Test thumbProps={{ "aria-label": "Label" }} />);
    rerender(<Test thumbProps={{ "aria-labelledby": "some-id" }} />);

    expect(error).not.toBeCalled();
    rerender(<Test />);
    expect(error).toBeCalledTimes(1);

    error.mockRestore();
  });

  describe("keyboard behavior", () => {
    it("should update the value correctly with specific keyboard keys", () => {
      const { getByRole } = render(<Test />);
      const slider = getByRole("slider");

      expect(slider).toHaveAttribute("aria-valuenow", "0");

      slider.focus();
      fireEvent.keyDown(slider, { key: "ArrowRight" });
      expect(slider).toHaveAttribute("aria-valuenow", "1");

      fireEvent.keyDown(slider, { key: "PageUp" });
      expect(slider).toHaveAttribute("aria-valuenow", "11");

      fireEvent.keyDown(slider, { key: "ArrowLeft" });
      expect(slider).toHaveAttribute("aria-valuenow", "10");

      fireEvent.keyDown(slider, { key: "ArrowLeft" });
      expect(slider).toHaveAttribute("aria-valuenow", "9");

      fireEvent.keyDown(slider, { key: "PageDown" });
      expect(slider).toHaveAttribute("aria-valuenow", "0");

      fireEvent.keyDown(slider, { key: "End" });
      expect(slider).toHaveAttribute("aria-valuenow", "100");

      fireEvent.keyDown(slider, { key: "Home" });
      expect(slider).toHaveAttribute("aria-valuenow", "0");
    });
  });
});
