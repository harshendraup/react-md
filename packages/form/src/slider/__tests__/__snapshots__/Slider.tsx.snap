// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider should render correctly 1`] = `
<div>
  <span
    class="rmd-label"
    id="slider-label"
  >
    Volume
  </span>
  <div
    class="rmd-slider-container rmd-slider-container--h rmd-slider-container--pad-left rmd-slider-container--pad-right"
  >
    <span
      class="rmd-slider-track rmd-slider-track--animate rmd-slider-track--hoverable rmd-slider-track--h rmd-slider-track--h1"
      id="slider"
      style="--offset1: 0%;"
    >
      <span
        aria-labelledby="slider-label"
        aria-valuemax="100"
        aria-valuemin="0"
        aria-valuenow="0"
        class="rmd-slider-thumb rmd-slider-thumb--h rmd-slider-thumb--h1 rmd-slider-thumb--animate"
        id="slider-thumb-1"
        role="slider"
        tabindex="0"
      />
      <input
        id="slider-thumb-1-value"
        type="hidden"
        value="0"
      />
    </span>
  </div>
</div>
`;

exports[`Slider should render correctly 2`] = `
<div>
  <span
    class="rmd-label"
    id="slider-label"
  >
    Volume
  </span>
  <div
    class="rmd-slider-container rmd-slider-container--h rmd-slider-container--pad-left rmd-slider-container--pad-right"
  >
    <span
      class="rmd-slider-track rmd-slider-track--animate rmd-slider-track--hoverable rmd-slider-track--h rmd-slider-track--h1"
      id="slider"
      style="--offset1: 0%;"
    >
      <span
        aria-labelledby="slider-label"
        aria-valuemax="100"
        aria-valuemin="0"
        aria-valuenow="0"
        class="rmd-slider-thumb rmd-slider-thumb--h rmd-slider-thumb--h1 rmd-slider-thumb--animate"
        id="slider-thumb-1"
        role="slider"
        tabindex="0"
      />
      <input
        id="slider-thumb-1-value"
        type="hidden"
        value="0"
      />
    </span>
  </div>
</div>
`;

exports[`Slider should render correctly 3`] = `
<div>
  <span
    class="rmd-label"
    id="slider-label"
  >
    Volume
  </span>
  <div
    class="rmd-slider-container rmd-slider-container--h rmd-slider-container--pad-left rmd-slider-container--pad-right"
  >
    <span
      class="rmd-slider-track rmd-slider-track--animate rmd-slider-track--hoverable rmd-slider-track--h rmd-slider-track--h1"
      id="slider"
      style="--offset1: 0%;"
    >
      <span
        aria-labelledby="slider-label"
        aria-valuemax="100"
        aria-valuemin="0"
        aria-valuenow="0"
        class="rmd-slider-thumb rmd-slider-thumb--h rmd-slider-thumb--h1 rmd-slider-thumb--animate"
        id="slider-thumb-1"
        role="slider"
        tabindex="0"
      />
      <input
        id="slider-thumb-1-value"
        type="hidden"
        value="0"
      />
    </span>
  </div>
</div>
`;

exports[`Slider should render correctly 4`] = `
<div>
  <span
    class="rmd-label"
    id="slider-label"
  >
    Volume
  </span>
  <div
    class="rmd-slider-container rmd-slider-container--pad-bottom rmd-slider-container--pad-top rmd-slider-container--v"
  >
    <span
      class="rmd-slider-track rmd-slider-track--animate rmd-slider-track--hoverable rmd-slider-track--v rmd-slider-track--v1"
      id="slider"
      style="--offset1: 0%;"
    >
      <span
        aria-labelledby="slider-label"
        aria-orientation="vertical"
        aria-valuemax="100"
        aria-valuemin="0"
        aria-valuenow="0"
        class="rmd-slider-thumb rmd-slider-thumb--v rmd-slider-thumb--v1 rmd-slider-thumb--animate"
        id="slider-thumb-1"
        role="slider"
        tabindex="0"
      />
      <input
        id="slider-thumb-1-value"
        type="hidden"
        value="0"
      />
    </span>
  </div>
</div>
`;

exports[`Slider should render correctly 5`] = `
<div>
  <span
    class="rmd-label"
    id="slider-label"
  >
    Volume
  </span>
  <div
    class="rmd-slider-container rmd-slider-container--pad-bottom rmd-slider-container--pad-top rmd-slider-container--v"
  >
    <span
      class="rmd-slider-track rmd-slider-track--animate rmd-slider-track--hoverable rmd-slider-track--v rmd-slider-track--v1"
      id="slider"
      style="--offset1: 0%;"
    >
      <span
        aria-labelledby="slider-label"
        aria-orientation="vertical"
        aria-valuemax="100"
        aria-valuemin="0"
        aria-valuenow="0"
        class="rmd-slider-thumb rmd-slider-thumb--v rmd-slider-thumb--v1 rmd-slider-thumb--animate"
        id="slider-thumb-1"
        role="slider"
        tabindex="0"
      />
      <input
        id="slider-thumb-1-value"
        type="hidden"
        value="0"
      />
    </span>
  </div>
</div>
`;

exports[`Slider should render correctly 6`] = `
<div>
  <span
    class="rmd-label"
    id="slider-label"
  >
    Volume
  </span>
  <div
    class="rmd-slider-container rmd-slider-container--pad-bottom rmd-slider-container--pad-top rmd-slider-container--v"
  >
    <span
      class="rmd-slider-track rmd-slider-track--animate rmd-slider-track--hoverable rmd-slider-track--v rmd-slider-track--v1"
      id="slider"
      style="--offset1: 0%;"
    >
      <span
        aria-labelledby="slider-label"
        aria-orientation="vertical"
        aria-valuemax="100"
        aria-valuemin="0"
        aria-valuenow="0"
        class="rmd-slider-thumb rmd-slider-thumb--v rmd-slider-thumb--v1 rmd-slider-thumb--animate"
        id="slider-thumb-1"
        role="slider"
        tabindex="0"
      />
      <input
        id="slider-thumb-1-value"
        type="hidden"
        value="0"
      />
    </span>
  </div>
</div>
`;

exports[`Slider should render correctly 7`] = `
<div>
  <span
    class="rmd-label rmd-label--disabled"
    id="slider-label"
  >
    Volume
  </span>
  <div
    class="rmd-slider-container rmd-slider-container--h rmd-slider-container--pad-left rmd-slider-container--pad-right"
  >
    <span
      class="rmd-slider-track rmd-slider-track--animate rmd-slider-track--disabled rmd-slider-track--h rmd-slider-track--h1"
      id="slider"
      style="--offset1: 0%;"
    >
      <span
        class="rmd-slider-thumb rmd-slider-thumb--h rmd-slider-thumb--h1 rmd-slider-thumb--animate rmd-slider-thumb--disabled rmd-slider-thumb--mask rmd-slider-thumb--mask-h"
      />
      <span
        aria-disabled="true"
        aria-labelledby="slider-label"
        aria-valuemax="100"
        aria-valuemin="0"
        aria-valuenow="0"
        class="rmd-slider-thumb rmd-slider-thumb--h rmd-slider-thumb--h1 rmd-slider-thumb--animate rmd-slider-thumb--disabled rmd-slider-thumb--disabled-h"
        id="slider-thumb-1"
        role="slider"
        tabindex="-1"
      />
      <input
        id="slider-thumb-1-value"
        type="hidden"
        value="0"
      />
    </span>
  </div>
</div>
`;

exports[`Slider should render correctly 8`] = `
<div>
  <span
    class="rmd-label rmd-label--disabled"
    id="slider-label"
  >
    Volume
  </span>
  <div
    class="rmd-slider-container rmd-slider-container--pad-bottom rmd-slider-container--pad-top rmd-slider-container--v"
  >
    <span
      class="rmd-slider-track rmd-slider-track--animate rmd-slider-track--disabled rmd-slider-track--v rmd-slider-track--v1"
      id="slider"
      style="--offset1: 0%;"
    >
      <span
        class="rmd-slider-thumb rmd-slider-thumb--v rmd-slider-thumb--v1 rmd-slider-thumb--animate rmd-slider-thumb--disabled rmd-slider-thumb--mask rmd-slider-thumb--mask-v"
      />
      <span
        aria-disabled="true"
        aria-labelledby="slider-label"
        aria-orientation="vertical"
        aria-valuemax="100"
        aria-valuemin="0"
        aria-valuenow="0"
        class="rmd-slider-thumb rmd-slider-thumb--v rmd-slider-thumb--v1 rmd-slider-thumb--animate rmd-slider-thumb--disabled rmd-slider-thumb--disabled-v"
        id="slider-thumb-1"
        role="slider"
        tabindex="-1"
      />
      <input
        id="slider-thumb-1-value"
        type="hidden"
        value="0"
      />
    </span>
  </div>
</div>
`;
